libsnark/zk_proof_systems/plonk/plonk.tcc

```diff
@@ -0,0 +1,412 @@
+    polynomial<FieldT> &f_poly)
+{
+    assert(L.size() != 0);
+    assert(L.size() == L[0].size());
+    assert(L.size() == f_points.size());
+
+    size_t num_gates = L.size();
+
+    // initialize f_poly
+    std::fill(f_poly.begin(), f_poly.end(), FieldT(0));
+
+    // the num_gates components of f_poly. each components is
+    // a polynomial that is the multiplication of the i-th element of
+    // f_points to the i-th polynomial in the lagrange basis L[i]:
+    // f_poly_component[i] = f_points[i] * L[i]
+    for (size_t i = 0; i < num_gates; ++i) {
+        polynomial<FieldT> f_poly_component_i;
+        // represent the scalar f_points[i] as an all-zero vector with
+        // only the first element set to f_points[i]. this is done in
+        // order to use libfqfft multiplication function as a function
+        // for multiply vector by scalar.
+        std::vector<FieldT> f_points_coeff_i(num_gates, FieldT(0));
+        f_points_coeff_i[0] = f_points[i];
+        // f_poly_component[i] = f_points[i] * L[i]
+        libfqfft::_polynomial_multiplication<FieldT>(
+            f_poly_component_i, f_points_coeff_i, L[i]);
+        // accumulate the i-th component to f_poly
+        libfqfft::_polynomial_addition<FieldT>(
+            f_poly, f_poly, f_poly_component_i);
+    }
+}
+
+/// Compute the selector polynomials of the given circuit (also
+/// called here "q-polynomials"). See Sect. 8.1.  The matrix
+/// gates_matrix_transpose has 5 rows, each corresponding to the
+/// values L, R, M, O and C for each gate; the number of columns is
+/// equal to the number of gates. L_basis is the Lagrange basis.
+template<typename FieldT>
+void plonk_compute_selector_polynomials(
+    const std::vector<std::vector<FieldT>> &gates_matrix_transpose,
+    std::vector<polynomial<FieldT>> &Q_polys)
+{
+    assert(gates_matrix_transpose.size() == Q_polys.size());
+    assert(gates_matrix_transpose[0].size() == Q_polys[0].size());
+    size_t num_qpolys = gates_matrix_transpose.size();
+    for (size_t i = 0; i < num_qpolys; ++i) {
+        std::vector<FieldT> q_vec = gates_matrix_transpose[i];
+        plonk_interpolate_polynomial_from_points<FieldT>(q_vec, Q_polys[i]);
+    }
+};
+
+template<typename FieldT>
+void plonk_compute_public_input_polynomial(
+    const std::vector<FieldT> &PI_points, polynomial<FieldT> &PI_poly)
+{
+    plonk_interpolate_polynomial_from_points<FieldT>(PI_points, PI_poly);
+};
+
+/// output: omega[0] are the n roots of unity, omega[1] are
+/// omega[0]*k1, omega[2] are omega[0]*k2; k1 H is a coset of H with
+/// generator omega_k1 distinct from H; k2 H is a coset of H with
+/// generator omega_k2, distinct from H and k1 H
+template<typename FieldT>
+void plonk_compute_roots_of_unity_omega(
+    const FieldT k1, const FieldT k2, std::vector<std::vector<FieldT>> &omega)
+{
+    assert(omega.size() == NUM_HSETS);
+    assert(omega[0].size() > 0);
+    size_t num_gates = omega[0].size();
```

Comment:
There could be a lot of room for error if this is passed in implicitly in the length of the first array.  Should it also be checked to ensure it's a power of 2?  Also that it's less than `2^(ppT::s)` (I think it's `s`)

Reply:
Potentially this and `plonk_roots_of_unity_omega_to_subgroup_H` can be replaced by a single function which takes an $n$, expected to be a power of 2, checks the validity of $n$ for the current curve, and returns the "H_gen" vector.  Then the caller doesn't have to be involved in the allocation of the vector.

Reply:
Proposed change accepted and applied. More specifically, replaced function `plonk_roots_of_unity_omega_to_subgroup_H` with `plonk_multiplicative_subgroups_H_k1H_k2H,` which recomputes the vector `omega` in order to avoid passing it as an input parameter potentially with the wrong size (as was done in `plonk_roots_of_unity_omega_to_subgroup_H`). Note that the function `plonk_compute_roots_of_unity_omega` is still kept (and used in `plonk_multiplicative_subgroups_H_k1H_k2H`) since (as mentioned in a reply to an earlier review comment) the `omega` vector is used in multiple places (e.g. as the `omega_roots` member of the `srs` class and the `circuit_t` structure).